### ISSUE TITLE
Standardize process for exchange channel(s) subscription

### ISSUE THREAD

**Issue by lukaszaoralek** [bitrich-info/xchange-stream#34]
@dozd 
Please consider the following proposal for subscribing process:
Now the subscription of the product(s)/symbol(s)/pair(s) is related with the methods like getOrderBook, getTrade and getTicker.
But for example the Gdax exchange allows only one subscription per multiple products. Another example is the Gemini exchange, where each product has own wss connection. In The bitfinex scenario, the channel request contains name of the product, but the server sends its channel id  in the response and then the server sends the market update/snapshot events with the channel id. In summary, each implementation has "tricky" code how to cover all of these cases.

In my opinion, it should be more standardize way how to subscribe the channles/product(s) without the "tricky" code. Therefore I propose the subscriptions of all product(s) and kind of market events have to be defined before connection to the exchange. For example:
In the interface StreamingExchange extends the method connect with one subscription argument:
```
public interface StreamingExchange<T extends StreamingSubscription> ... {
  ...
  Completable connect(T streamingSubscription);
  ...
}
```

Each exchange has to be implemented the own class (ExchangeName)StreamingSubscription (that implements interface StreamingSubscription). Consider the example with Gdax and Gemini cases:
```
class GdaxStreamingSubscription implements StreamingSubscription {
  private GdaxStreamingSubscription() {...}

  public static createGdaxStreamingSubscription(GdaxProductSubscription[] subscriptions) {}
}
```

where the class GdaxProductSubscription can look like:
```
GdaxProductSubscription {
  public GdaxProductSubscription(CurrencyPair pair, boolean orderbook, boolean ticker, boolean trade) {...}
}
```

and for Gemini case the StreamingSubscription class will be different:
```
class GeminiStreamingSubscription implements StreamingSubscription {
  private GeminiStreamingSubscription() {...}
  ...
  public static createGeminiStreamingSubscription(CurrencyPair[] pairs) {}
  ...
}
```

The GdaxExchange will look like:
```
public class GDAXStreamingExchange extends GDAXExchange implements StreamingExchange<GdaxStreamingSubscription> {
  ...
  public Completable connect(GdaxStreamingSubscription streamingSubscription) {...}
  ...
}
```

It hides all needs for subscribing the market events for different exchanges, standardize the subscribing process, and the StreamingMarketDataService interface doesn't change.

What do you think guys?

**Comment by dozd** [comment]
Like you mean creating all subscriptions during connect (based on exchange specifics) and then just make observable streams through `StreamingMarketDataService` for already subscribed products, am I right?

One thing which I'm not really comfortable with is that if you don't put right bool into `StreamingSubscription` than calling `StreamingMarketDataService.getOrderBook` is not working. It's like there is some hidden argument dependency in the interface (you must call this method with right arguments to have next method working). But that could be solved with some design iterations I think.

**Comment by lukaszaoralek** [comment]
@dozd 
Yeah, you're right.

I'm not sure I get the point of your second statement. Let's say we want to subscribe market events from GDAX. Concretely, we want to obtain orderbook snapshots for BTC/USD and trades for ETH/USD. First of all we prepare subscription request for GDAX exchange:
```
GdaxProductSubscription[] products = new GdaxProductSubscription[] {
new GdaxProductSubscription(CurrencyPair.BTC_USD, true, false, false),
new GdaxProductSubscription(CurrencyPair.ETH_USD, false, false, true)};
GdaxStreamingSubscription gdaxSubscription = GdaxStreamingSubscription.createGdaxStreamingSubscription(gdaxProducts);
```
Then we connect to the exchange:
```
GDAXStreamingExchange gdax = new GDAXStreamingExchange();
gdax.connect(gdaxSubscription).allowBlocking();
...
```
and now we can subscribe orderbook snapshot for BTC/USD and trades for ETH/USD like:
```
StreamingMarketDataService marketDataService = gdax.getStreamingMarketDataService();
marketDataService.getOrderBook(CurrencyPair.BTC_USD).subscribe(e -> {...});
marketDataService.getTrades(CurrencyPair.ETH_USD).subscribe(e -> {...});
```
If we add also "bad subscription" orderbook and trades for CurrencyPair.LTC_USD and BTC_USD, respectively:
```
marketDataService.getOrderBook(CurrencyPair.LTC_USD).subscribe(e -> {...});
marketDataService.getTrades(CurrencyPair.BTC_USD).subscribe(e -> {...});
```
simply nothing happen, we don't get any market events from that combination. Of course it is possible to throw exception for wrong subscription.

It's like getOrderBook, getTicker and getTrades methods just filtrate incoming market events. Because right now, these methods are also responsible for channel requests. For some exchanges like GDAX it is problem, because channel requests are part of one request. I would like to separate two things: connection (including channel requests) and subscription of market events.

**Comment by dozd** [comment]
I'm not 100% sure if you can get rid of "tricky code" you are writing about. Don't you just move it into different place?  Either way you need to deal with it (either during connect or later). I get the GDAX problem where there is only one subscription because current design is like there is a subscribe command 1:1 to data subscription.

So it's obvious that there is a flaw in the current design.

I need to think about it a little more because for example class `GdaxProductSubscription` seems odd to me on the first look. It should have better contract (e.g. not using boolean arguments, use builder pattern maybe, dunno now). There should be adding / deleting subscription also. Maybe mixture of current design (for 1:1 subscribe commands : data) with some additions to have better ways how to implement different scenarios.

**Comment by dozd** [comment]
I'd like to have "user contract" for library users as simple as possible. So don't take my stance too negative, I'd like to think something that has smallest user impact possible :). Because user experience >> a lot of tricky code in exchange implementation (if it's possible).

**Comment by lukaszaoralek** [comment]
@dozd 
I understand you. I don't take your stance too negative, don't worry. :) It's just discussion about the subscription design. The GdaxProductSubscription was an example, of course there can be different mechanism how to subscribe concrete currency and market events.

I think the main question now is if you basically agree or disagree with subscription of products during connection, somethink like:
`Completable connect(T streamingSubscription);`
Because in this scenario, I'll know everything what the user want to subscribe and I can make/construct correct connection and channel requests without 'tricky code'. From the user point of view it means the channel request(s) and connection will be really during connection process; it just move from getTrades/Orderbook/Ticker methods to connection method, nothing more. I think the library will remain as simple as now. :)

Do you mean it is bad idea in general?

**Comment by dozd** [comment]
Not in general. I'm not just comfortable that you must say: i want this in the connect and then again with call for market service - it seems redundant in some scenarios. But if it's inevitable for some exchanges, there should be a way (like overloaded variant of connect). I wouldn't require it for all exchanges if current interface is sufficient. I like the laziness of the approach. You are allowed to change the subscriptions on the fly.

The question is, what streaming subscription in `connect(StreamingSubscription) should look like?
- Is there a need for exchange specific subscriptions (`GdaxProductSubscription`)?
- Can we have only one for all exchanges?
- Can we use builder pattern with fluent interface e.g. `Subscription.create().add(Ticker, BTC/USD).add(Ticker, ETH/USD).add(OrderBook, BTC/USD).build();`

Another questions I have in mind
- If the connect is called with subscription object - how is user allowed to subscribe for another product?
- Will there be an exception when user request data from makret service that he did not subscribed?

**Comment by lukaszaoralek** [comment]
Ok, based on your questions and needs what about following design:

```
public class SubscriptionArgument {
  private SubscriptionArgument() {}
}

public interface StreamingExchange<T extends SubscriptionArgument> {
  ...
  Completable connect(T... args);
  ...
}

class BitstampStreamingExchange extends StreamingExchange<SubscriptionArgument> {
  Completable connect(SubscriptionArgument... args);
  ...
}

class GDAXStreamingExchange extends StreamingExchange<GDAXSubscriptionArgument> {
  Completable connect(GDAXSubscriptionArgument... args);
  ...
}

public class GDAXSubscriptionArgumentBuilder extends SubscriptionArgument {
  private List<CurrencyPair> orderbook;
  private List<CurrencyPair> trades;
  private List<CurrencyPair> ticker;

  private GDAXSubscriptionArgument(GDAXSubscriptionArgumentBuilder builder)
    ...
  }

  public static class GDAXSubscriptionArgumentBuilder {
    ...
    private GDAXSubscriptionArgumentBuilder() {
      orderbook = new ArrayList<>();
      trades = new ArrayList<>();
      ticker = new ArrayList<>();
    }

    public GDAXSubscriptionArgumentBuilder addOrderbook(CurrencyPair pair) {
      orderbook.add(pair);
      return this;
    }
    ...

    public GDAXSubscriptionArgument build() {
      return new GDAXSubscriptionArgument(this);
    }
  }

  public static GDAXSubscriptionArgumentBuilder create() {
    return new GDAXSubscriptionArgumentBuilder();
  }
}
```
Now the instantiation of the exchanges can be like that:
```
CurrencyPair btcUsd = new CurrencyPair.BTC_USD;
CurrencyPair ethUsd = new CurrencyPair.ETH_USD;
StreamingExchange<GDAXSubscriptionArgument> gdax = StreamingExchangeFactory.INSTANCE.createExchange(GDAXStreamingExchange.class.getName());
GDAXSubscriptionArgument arg = GDAXSubscriptionArgument.create()
  .addOrderbook(btcUsd).addOrderbook(ethUsd).build();
gdax.connect(arg).blockingAwait();

StreamingExchange<SubscriptionArgument> bitstamp = StreamingExchangeFactory.INSTANCE.createExchange(BitstampStreamingExchange.class.getName());
bitstamp.connect().blockingAwait();
```
In this scenario the Bitstamp exchange has connect method without arguments just like now (the class SubscriptionArgument has private constructor), but the GDAX exchange has mandatory  GDAXSubscriptionArgument with builder pattern.

If the user requests market events from the market service that the user did not subscribe (only for the GDAX exchange or similar), the market service throws the exception like ticker subscription from bitstamp market service right now.

What do you think?

**Comment by dozd** [comment]
Yep, but `GDAXSubscriptionArgument` doesn't need to be GDAX specific, does it? I think it can be same for all exchanges (that needs it in connect).

**Comment by lukaszaoralek** [comment]
I thought each exchange that needs the SubscriptionArgument, extends own version of the SubscriptionArgument, because what the GDAX needs in SubscriptionArgument  can be different with other exchanges. Or do you mean something else?

**Comment by dozd** [comment]
Nope. But how different? Isn't product definition same for all exchanges? Trade / Ticker / Orderbook with currency pair?

**Comment by dozd** [comment]
I wouldn't reuse `SubscriptionArgument` for two different things - one is product selection and second for example mentioned private API setup. That could be confusing.

As of private API, I'd go same way as Xchange library is going - we can reuse their code for settings private keys (i think it's done through ExchangeSpecification which can be available also for xchange-stream). Don't know but your keys will be probably same for streaming API as well as rest api. So user using private REST api can use streaming api right without any additional setup.

I can't think of another usage of `SubscriptionArgument` right now so I'd like to see simplest way of implementation - no abstraction, no inheritance, just one interface / implementation to rule them all :).

**Comment by lukaszaoralek** [comment]
I see ...so, it will be just (Product)SubscriptionArgument as only one class, correct?

**Comment by dozd** [comment]
Cool, `ProductSubscription` will be sufficient. Argument postfix is
redundant.

**Comment by lukaszaoralek** [comment]
Ok :)

**Comment by lukaszaoralek** [comment]
@dozd 
PR #38
